File: leetcode2/1easy/최원준/Q892.java
package Leetcode.최원준;

/*
1. 아이디어 :


2. 시간복잡도 :
O( )

3. 자료구조/알고리즘 :

 */

public class Q892 {
    class Solution {
        int[] dx = {0,0,1,-1}, dy = {1,-1,0,0};
        int n,m ;
        int[][] grid;
        public int countArea(int row, int col) {
            int height = grid[row][col];
            int area = (height == 0) ? 0 : 2; //bottom, top

            for (int i=0; i<4; i++) {
                int nx = row + dx[i], ny = col + dy[i];
                int neighbor = (nx<0 || ny<0 || nx>=n || ny>=m) ? 0 : grid[nx][ny];
                area += Math.max(0, height - neighbor);
            }
            return area;
        }

        public int surfaceArea(int[][] grid) {
            n = grid.length;
            m = grid[0].length;
            this.grid = grid;

            int ans = 0;
            for (int i=0; i<n; i++) for (int j=0; j<m; j++) {
                ans += countArea(i,j);
            }
            return ans;
        }
    }
}

File: leetcode2/1easy/강병호/Q892.py
class Leetcode.이재훈.이재훈.Solution:
    def surfaceArea(self, grid: List[List[int]]) -> int:
        total = 0
        
        for i in range(len(grid)):
            for j in range(len(grid)):
                if grid[i][j] > 0:
                    total += 2 + grid[i][j] * 4
                    
                    if i > 0:
                        total -= min(grid[i][j], grid[i-1][j]) * 2
                        
                    if j > 0:
                        total -= min(grid[i][j], grid[i][j-1]) * 2
                        
        return total
